lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+		txnYieldEarningDurationNanoSecs = leftYieldCurvePoint.LockupDurationNanoSecs
+	}
+	if rightYieldCurvePoint.LockupDurationNanoSecs == lockupDurationNanoSeconds {
+		txnYieldBasisPoints = rightYieldCurvePoint.LockupYieldAPYBasisPoints
+		txnYieldEarningDurationNanoSecs = rightYieldCurvePoint.LockupDurationNanoSecs
+	}
+
+	// Convert variables to a consistent uint256 representation. This is to use them in SafeUint256 math.
+	txnYieldBasisPoints256 := uint256.NewInt().SetUint64(txnYieldBasisPoints)
+	txnYieldEarningDurationNanoSecs256 := uint256.NewInt().SetUint64(uint64(txnYieldEarningDurationNanoSecs))
+
+	// Compute the yield associated with this operation, checking to ensure there's no overflow.
+	yieldFromTxn, err :=
+		CalculateLockupYield(txMeta.LockupAmountBaseUnits, txnYieldBasisPoints256, txnYieldEarningDurationNanoSecs256)
+	if err != nil {
+		return 0, 0, nil, errors.Wrap(err, "_connectCoinLockup")
+	}
+
+	// We check that the minted yield does not cause an overflow in the transactor's balance.
+	// In the case of DeSo being locked up, we must check that the resulting amount is less than 2**64.
+	if uint256.NewInt().Sub(MaxUint256, yieldFromTxn).Lt(transactorBalanceNanos256) {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupYieldCausesOverflow, "_connectCoinLockup")
+	}
+
+	// Compute the amount to be added to the locked balance entry.
+	lockupValue := *uint256.NewInt().Add(transactorBalanceNanos256, yieldFromTxn)
+
+	// In the case of DeSo being locked up, we ensure that the resulting amount is less than 2**64.
+	if profilePKID.IsZeroPKID() && !lockupValue.IsUint64() {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupYieldCausesOverflow, "_connectCoinLockup")
+	}
+
+	// NOTE: While we could check for "global" overflow here, we let this occur on the unlock transaction instead.
+	//       Global overflow is where the yield causes fields like CoinEntry.CoinsInCirculationNanos to overflow.
+	//       Performing the check here would be redundant and may lead to worse UX in the case of coins being
+	//       burned in the future making current lockups no longer an overflow. Checking here would also
+	//       create a DoS attack vector where a malicious entity takes out an extremely long-dated lockup
+	//       with the sole intent of saturating the CoinsInCirculationNanos field preventing others from locking up.
+
+	// For consolidation, we fetch equivalent LockedBalanceEntries.
+	lockedBalanceEntry := bav.GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+		hodlerPKID, profilePKID, txMeta.UnlockTimestampNanoSecs)
+	if lockedBalanceEntry == nil || lockedBalanceEntry.isDeleted {
+		lockedBalanceEntry = &LockedBalanceEntry{
+			HODLerPKID:                  hodlerPKID,
+			ProfilePKID:                 profilePKID,
+			ExpirationTimestampNanoSecs: txMeta.UnlockTimestampNanoSecs,
+			BalanceBaseUnits:            *uint256.NewInt(),
+		}
+	}
+	previousLockedBalanceEntry := *lockedBalanceEntry
```

Comment:
I don't see a bug with this right now, but it makes me a little uncomfortable that we allow the user to modify an existing LockedBalanceEntry. It works here because you only allow them to ADD to it. I think this is OK. It would be a little safer to just disallow modifying existing entries, thus forcing them to create a new one that's like 1 nanosecond in the future. But if you can just inspect this really carefully to make sure there are no bugs with the modify approach, I think we can keep it the way it is.

Reply:
While I agree, I think it's a bit annoying given that we have users specify the lockup expiration in their lockup transaction. Having to put a lockup 1 nanosecond in the future would be a pretty bad UX, especially if we were supporting something like locked coin distributions.